packages/plugin-hardhat/src/utils/deploy.ts

```diff
@@ -1,8 +1,18 @@
+import type { ContractFactory, ContractTransaction } from 'ethers';
 import type { Deployment } from '@openzeppelin/upgrades-core';
-import type { ContractFactory } from 'ethers';
+export interface DeploymentExecutor {
+  (factory: ContractFactory): Promise<HardhatDeployment>;
```

Comment:
So I can't exactly recall why this was necessary, but I believe it had something to do with the transformation of the proxy deployment from 
```ts
const proxy = await ProxyFactory.deploy(impl, adminAddress, data);
```
into 
```ts
const proxy = await deploy(ProxyFactory, [impl, adminAddress, data]);
 ```
 and its return values not matching up. Notice how we don't use `intoCoreDeployment` on this particular line.

Reply:
Ahh, yes, I believe I recall why we needed `HardhatDeployment` here is because at the time we pass the function onto the deployer, we do not actually have the transaction hash and the functionality of your code base requires a provider in order to fetch transaction by hash... Sorry if this isn't entirely clear just yet, but I will try again and see if I can remind myself exactly why this workaround was introduced.

Reply:
I'd imagine the `deploy` function that was originally in this file would be a good `DeploymentExecutor`, where the hash is returned instead of the full `ContractTransaction`.

Reply:
The reason I insist on this is that we currently have duplication of this layer across the Hardhat and Truffle plugins but we will want to merge those soon, and using the transaction hash is a more agnostic way to represent a deployment, instead of the Ethers-specific `ContractTransaction`.

Reply:
Totally fine with me. I prefer to return only the hash any way. Have made the necessary adjustments and it also made the code on our end much more attractive and easy to write.